tf.1
.TH TF(1)
.SH NAME
tf - launch a container on 20ft.nz
.SH SYNOPSIS
\fBtf\fR \-h | \fBtf\fR --help

\fBtf\fR [option]... source

\fBtf\fR --systemd user@my.server.com [option]... source


.SH DESCRIPTION
.B tf
launches containers or scripts on 20ft.nz, passes environment variables, selects a starting command, connects stdio, copies pre-boot files, mounts persistent volumes, port maps onto the container, and publishes to the web. For developers it can create an ssh/sftp server, and write a systemd unit file to a server.

The source is assumed to be a docker id in either the 'hex' or 'name' form; or '.' implies the most recently added image in the local docker daemon.

.SH OPTIONS
.TP
.BR \-h ", " \-\-help
Present cli help.
.TP
.BR \-\-location\ \fRx\.20ft\.nz
Use an (optional) non-default location (fqdn).
.TP
.BR \-\-local\fR\ y.local
An optional (local) ip for the broker.

.TP
.BR \-v,\ --verbose
Verbose logging (level DEBUG).
.TP
.BR \-q,\ --quiet
Logging is not configured.

.TP
.BR \-i,\ --interactive
Run interactively. This connects stdin and stdout to the terminal and provides an escape sequence (triple ^]).
.TP
.BR \-e,\ --environment\fR\ ENV=value
Add an environment variable to the launch context.
.TP
.BR \-f,\ --file\fR\ src:dest
Before boot copy the source file into the destination directory.
.TP
.BR \-m,\ --mount\fR\ tag:/mount/point
Mount, at /mount/point, the volume given by either a uuid, tag, or uuid:tag pair.
.TP
.BR \-p,\ --publish\fR\ localport:remoteport
Create a TCP tunnel between localhost and the container on the given local and destination ports. Note that local ports < 1024 will need superuser privileges.
.TP
.BR \-c,\ --command,\ --entrypoint\fR\ /some/path
Choose an alternate command/entrypoint into the container.
.TP
.BR \-w,\ --web\fR\ [subdomain.]wip.my.com[:www.my.com[:certname]]
Publish port 80 of the container to a web endpoint. Publishing to a subdomain is enabled by creating a wildcard DNS entry (i.e. *.wip). Applications which expect a given 'Host' header can have this set with the rewrite (second) option. To host behind ssl pass a certificate name where tf will look for certname.crt and certname.key. Intermediate certs will need to be cat'ed onto the end of the main certificate.

.TP
.BR \-s,\ \-\-ssh\fR\ port
Launch the container, then start a shell and wrap it with an ssh session on the specified port (pass:root, blank password).
.TP
.BR \-z,\ \-\-sleep
Launch the container but don't run the entrypoint or command.

.TP
.BR \-\-systemd\fR\ user@server.my.com
Write a systemd unit on a server instead of running tf locally. For an image tagged myco/image it will create a subdirectory of the user's home directory called myco-image and place any necessary resources in there; then create and start a systemd service called myco-image with sensible defaults.

.TP
.BR \-\-identity\fR\ ~/.ssh/some_id.pem
Use the given identity file to connect via ssh.

.SH EXAMPLES
.PP
Launch the most recently built Docker image and map local port 8080 to remote port 80.
.PP
.RS
tf -p 8080:80 .
.RE

.PP
Launch the stock nginx image mapping port 8080 to 80, with a pre-existing volume for the http content and the ability to ssh/sftp into the container.
.PP
.RS
tf -p 8080:80 -m persistent_web:/usr/share/nginx/html --ssh 2222 nginx
.RE

.PP
Launch a docker image (possibly only held locally), map to web endpoint, rewrite the http-request header to client.com, and use an ssl certificate given by the combination of wip.crt and wip.key.
.PP
.RS
tf -w wip.my.com:client.com:wip my/image
.RE

.PP
Create a systemd unit to run the above operation as a service on a server (noting that the web traffic will still be served directly from the 20ft cluster).
.PP
.RS
tf --systemd -w wip.my.com:client.com:wip my/image
.RE


tfnz/cli/tf.py
# Copyright (c) 2017 David Preece, All rights reserved.
#
# Permission to use, copy, modify, and/or distribute this software for any
# purpose with or without fee is hereby granted.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.


# the method to implement is tf_main(location, node, environment, portmap, pass_args)

import sys
import re
import os
import os.path
import argparse
import signal
from sys import argv, stderr
from importlib import import_module
from messidge import default_location, create_account
from tfnz.location import Location
from tfnz.volume import Volume
from tfnz.docker import Docker
from tfnz.endpoint import Cluster
from tfnz.cli import base_argparse, systemd, Interactive


def main_impl():
    # see if we have an account
    try:
        default_location(prefix="~/.20ft")
    except RuntimeError:
        # there is no default location
        if len(argv) != 3:
            print("""
Either there is no a 20ft account on this machine, and you need 
to request access from your administrator; or the ~/.20ft/default_location
file has been deleted. Run something like...

    echo 'location.20ft.nz' > ~/.20ft/default_location

If you already have an account on another machine you can merely 
copy the directory ~/.20ft (and it's contents) to this machine.""", file=stderr)
            return None

        # trying to create a new account using token passed by the server
        create_account(argv[1], argv[2], prefix="~/.20ft")
        print("Created OK for: %s\n" % argv[1])
        return None

    # right, get on with it
    parser = base_argparse('tf')
    launch_group = parser.add_argument_group('launch options')
    launch_group.add_argument('-v', '--verbose', help='verbose logging', action='store_true')
    launch_group.add_argument('-q', '--quiet', help='no logging', action='store_true')
    launch_group.add_argument('-ti', '--interactive', help='interactive, connect stdin and stdout', action='store_true')
    launch_group.add_argument('-e', '--env', help='set an environment variable, possibly from current',
                              action='append', metavar='[VAR=value | VAR]')
    launch_group.add_argument('-f', '--file', help='write a pre-boot file', action='append', metavar='src:dest')
    launch_group.add_argument('-m', '--mount', help='mount a volume', action='append',
                              metavar='[uuid:mountpoint | tag:mountpoint]')
    launch_group.add_argument('-p', '--publish', help='add a local->remote tcp proxy', action='append',
                              metavar='8080:80')
    launch_group.add_argument('-w', '--web', help='publish on web endpoint',
                              metavar='subdomain.my.com[:www.my.com[:certname]]')
    development_group = parser.add_argument_group('development options')
    development_group.add_argument('--ssh', help='create an ssh/sftp wrapped shell on given port',
                                   metavar='2222')
    development_group.add_argument('-s', help='shorthand for --ssh 2222', action='store_true')
    development_group.add_argument('-z', '--sleep', help='launch the container asleep (instead of entrypoint)',
                                   action='store_true')
    server_group = parser.add_argument_group('server options')
    server_group.add_argument('--systemd', help='create a systemd service', metavar='user@server.my.com')
    server_group.add_argument('--identity', help='specify an identity file to use with --systemd',
                              metavar="~/.ssh/some_id.pem")

    parser.add_argument('source', help="if '.' runs the most recently added docker image; "
                                       "if ends in '.py' tries to run a function implementation; "
                                       "else this is the tag or hex id of an image to run.")

    parser.add_argument('command', help='run this command/entrypoint instead', nargs='?')
    parser.add_argument('args', help='arguments to pass to a script or subprocess', nargs=argparse.REMAINDER)

    args = parser.parse_args()
    function_implementation = len(args.source) > 3 and args.source[-3:] == ".py"

    # collect any pre-boot files
    preboot = []
    if args.file is not None:
        for e in args.file:
            match = re.match('^[\w:]*[\w.\-\\/]+:[\w.\-\\/]+$', e)
            if not match:
                print("Pre-boot copies need to be in source:destination pairs", file=sys.stderr)
                print("....error in '%s'" % e, file=sys.stderr)
                return None
            files = e.split(':')
            # are we copying into a directory? append the original filename
            if files[1][-1:] == '/':
                files[1] += files[0]
            # try to read the file in
            try:
                with open(files[0], 'rb') as f:
                    preboot.append((files[1], f.read()))
            except FileNotFoundError:
                print("Could not find the source pre-boot file: " + files[0], file=sys.stderr)
                return None

    # if publishing to an endpoint, we might need ssl certs
    endpoint = None
    rewrite = None
    cert = None
    fqdns = None
    if args.web is not None:
        # split into endpoint:rewrite:certname (rewrite and certname are optional)
        fqdns = args.web.split(':')
        if len(fqdns) == 0 or len(fqdns[0]) == 0:
            print("Cannot publish to an endpoint without an address", file=sys.stderr)
            return None
        endpoint = fqdns[0]

        if len(fqdns) > 1 and len(fqdns[1]) > 0:
            rewrite = fqdns[1]

        if len(fqdns) > 2 and len(fqdns[2]) > 0:
            cert = (fqdns[2] + '.crt', fqdns[2] + '.key')
            if not os.path.exists(cert[0]):
                print("Cannot find certificate for ssl: " + cert[0], file=sys.stderr)
                return None
            if not os.path.exists(cert[1]):
                print("Cannot find key for ssl: " + cert[1], file=sys.stderr)
                return None

    # connect
    location = None
    try:
        location = Location(args.location, location_ip=args.local, quiet=args.quiet, debug_log=args.verbose)
    except BaseException as e:
        print("Failed while connecting to location: " + str(e), file=sys.stderr)
        return location

    # are we making a systemd service?
    if args.systemd is not None:
        systemd(location, args, argv, preboot, cert)
        return location

    # are we going to be using the most recent build?
    if args.source == '.':
        args.source = Docker.last_image()

    # create env-vars
    e_vars = set()
    environment = []
    if args.env is not None:
        for e in args.env:
            match_single = re.match('^[[0-9A-Za-z:_]+$', e)
            if match_single:
                if e not in os.environ:
                    print("Environment variable \'%s\' was not found in the local environment.", file=sys.stderr)
                    return location
                e += '=\'%s\'' % os.environ[e]
            match_extended = re.match('^[0-9A-Za-z:_]+=', e)
            if not match_extended:
                print("Environment variables need to be passed as 'name' or 'name=value' pairs", file=sys.stderr)
                print("....error in '%s'" % e, file=sys.stderr)
                return location
            variable = match_extended.group(0)[:-1]
            if variable in e_vars:
                print("Can only pass one value per environment variable.", file=sys.stderr)
                print("....error in '%s'" % e, file=sys.stderr)
                return location
            value = e[len(variable)+1:]
            e_vars.add(variable)
            environment.append((variable, value))

    # create portmaps
    l_ports = set()
    portmap = []
    if args.publish is not None:
        for e in args.publish:
            match = re.match('\d+:\d+$', e)
            if not match:
                print("Portmaps need to be passed as number:number pairs", file=sys.stderr)
                print("....error in '%s'" % e, file=sys.stderr)
                return location
            local, remote = e.split(':')
            if local in l_ports:
                print("Cannot bind a local port twice.", file=sys.stderr)
                print("....error in '%s'" % e, file=sys.stderr)
                return location
            l_ports.add(local)
            portmap.append((local, remote))

    # create volume mounts ensuring they don't overlap
    volumes = []
    mount_points = set()
    if args.mount is not None:
        for m in args.mount:
            if ':' not in m:
                print("Volumes need to be passed as uuid:mountpoint pairs", file=sys.stderr)
                print("....error in '%s'" % m, file=sys.stderr)
                return location
            find = m.rfind(':')
            key = m[:find]
            mount = m[find+1:]
            intersection = Volume.trees_intersect(mount_points, mount)
            if intersection is not None:
                print("Error in volumes: %s is a subtree of %s" % (intersection[0], intersection[1]), file=sys.stderr)
                return location
            try:
                volumes.append((location.volumes.get(location.user_pk, key), mount))
            except KeyError:
                print("Could not find volume: " + key)
                return location
            mount_points.add(mount)

    # overloaded command? args? (makes it into a list)
    if args.command is not None and args.args is not None:
        args.command = [args.command]
        args.command.extend(args.args)

    # try to launch as a method implementation
    if function_implementation:
        try:
            sys.path.append(os.path.dirname(os.path.expanduser(args.source)))
            sys.path.append(os.getcwd())
            imp = import_module(os.path.basename(args.source[:-3]))
            imp.tf_main(location, environment, portmap, preboot, volumes, args.command, args.args)
            signal.pause()
        except (ImportError, TypeError, AttributeError):
            print("Failed to import or run function in: " + args.source, file=sys.stderr)
            print("The function to implement is: "
                  "def tf_main(location, environment, portmap, preboot, volumes, cmd, args):",
                  file=sys.stderr)
            return location
        except KeyboardInterrupt:
            return location

    # try to launch the container
    interactive = Interactive(location) if args.interactive else None
    try:
        node = location.node()
        container = node.spawn_container(args.source,
                                         env=environment,
                                         pre_boot_files=preboot,
                                         volumes=volumes,
                                         stdout_callback=(interactive.stdout_callback if args.interactive
                                                          else lambda _, out: sys.stdout.buffer.write(out)),
                                         termination_callback=(interactive.termination_callback if args.interactive
                                                               else location.disconnect),
                                         command=args.command,
                                         sleep=args.sleep)
        container.wait_until_ready()  # a transport for exceptions
    except BaseException as e:
        print("Failed while spawning container: " + str(e), file=sys.stderr)
        return location

    # create the tunnels
    for m in portmap:
        container.attach_tunnel(m[1], m[0])

    # launch an ssh server? user/pass are anything/anything - you can only connect from localhost anyway
    if args.ssh or args.s:
        container.create_ssh_server(2222 if args.s else int(args.ssh))

    # publish to an endpoint?
    if args.web is not None:
        container.wait_until_ready()
        clstr = Cluster([container], rewrite=rewrite)
        ep = location.endpoint_for(endpoint)
        ep.publish(clstr, fqdns[0], ssl=cert)

    # wait until quit
    if args.interactive:
        interactive.stdin_loop(container)
    else:
        try:
            location.conn.wait_until_complete()
        except (KeyboardInterrupt, AttributeError):
            pass
    return location


def main():
    maybe_location = main_impl()
    if maybe_location is not None:
        maybe_location.disconnect()


if __name__ == "__main__":
    main()
